### PR TITLE
New `extra-files` field for neutral `sdist` files

### LINKED ISSUE #8817: Neutral field (like `extra-files` or `extra-sdist-files`) to add files to `sdist`
We currently have 3 ways to put extra files into the sdist tarball, but all of these have special semantics:
1. `data-files`: files used by the package at runtime
2. `extra-source-files`: files that may be used during `cabal build` (and thus changes might trigger a rebuild)
3. `extra-doc-files`: copied by `cabal haddock` into the documentation folder

Often one of the latter is abused to add files that have no semantics for `cabal`, like `stack*.yaml` files to build with `stack`, or files like READMEs or CHANGELOGs that maybe be regarded by Hackage, but are misplaced in any of the 3 classes.

Suggestion: add a fourth, neutral field for adding files without semantics, e.g.
```
extra-sdist-files:
  README.md
  CHANGELOG.md
  stack-9.4.4.yaml
  stack-9.2.6.yaml
  ...
```

Or, make a structured `extra-files` entry:
```
extra-files:
  source:
    config.mk  --whatever, I never use this
    ...
  doc:
    ... -- lacking imagination here as well
  other:
    CHANGELOG.md
    README.md
    stack-9.4.4.yaml
    ...
```

Lifted from:
- #8756

### PR DESCRIPTION
**Pull request by tbidne**
Resolves #8817.

This PR adds a new `extra-files` field that provides a way to specify extra files that should be included in `sdist`, without adding any other semantics (cf. `extra-source-files` are tracked by `cabal build`).

### Field name bikeshedding

I decided to go with `extra-files`. Alternatives considered:

- Structured e.g. from the issue:

    ```cabal
    extra-files:
      source:
        config.mk  --whatever, I never use this
        ...
      doc:
        ...
      other: -- this would be the new field
        ...
    ```
    
    I actually like this idea in principle, but of course we have to consider how this fits into the existing spec. I cannot imagine anyone has the appetite to replace the current top-level `extra-*-files`, and I don't think adding an alternative here makes sense just for adding a "neutral option".
    
- `extra-sdist-files`:

    I think this is the best sounding name that fits the `extra-<something>-files` format. The only downside is if there is a future possibility that such neutral extra files would ever be wanted for something other than `sdist`. It would be a shame if the answer to a question like
    
    > How do I add extra files for X?
    
    Had an answer of
    
    > Add them to `extra-sdist-files`.
    
    Thus my preference for something more general. Though maybe that future scenario would likely suggest a new `extra-?-files` option, so perhaps this fear isn't warranted.
    
- `extra-other-files`, `extra-misc-files`, etc.: I can't really come up with a great alternative here. `other` fits well with the structured approach above, but `extra-other-files` sounds weird, imo. I'm ambivalent on `extra-misc-files`.

Hence I went with `extra-files`. But I am pretty undecided on this, so please do offer suggestions.

### Other notes

- After writing the test in the `cabal-testsuite/PackageTests/SDist` directory, I noticed that there is also a `cabal-testsuite/PackageTests/NewSdist` dir. Just checking that it's in the right place.

- I added an entry in `file-format-changelog.rst`, though I'm unsure of the next cabal version (I presume this is not going in `3.12`). Maybe `3.14`?

- Probably there are other documentation changes needed (e.g. wiki), though I don't know exactly where those should be.

Thanks!

---

**Template Α: This PR modifies [behaviour or interface](https://github.com/cabalism/cabal/blob/master/CONTRIBUTING.md#changelog)**

Include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [x] The documentation has been updated, if necessary.
* [ ] [Manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) have been included.
* [x] Tests have been added. (*Ask for help if you don’t know how to write them! Ask for an exemption if tests are too complex for too little coverage!*)

### CODE REVIEW

**Comment by andreabedini** [comment]
Thanks! This requires a bump in `cabal-version`, also it needs to be discussed.

**Comment by tbidne** [comment]
I was wondering about that (3.14?). And indeed, happy to discuss :slightly_smiling_face: .

File: Cabal-syntax/src/Distribution/PackageDescription/FieldGrammar.hs [hunk 1]
```diff
@@ -143,6 +143,7 @@ packageDescriptionFieldGrammar =
     <*> monoidalFieldAla "extra-source-files" formatExtraSourceFiles L.extraSrcFiles
     <*> monoidalFieldAla "extra-tmp-files" (alaList' VCat RelativePathNT) L.extraTmpFiles
     <*> monoidalFieldAla "extra-doc-files" formatExtraSourceFiles L.extraDocFiles
+    <*> monoidalFieldAla "extra-files" formatExtraSourceFiles L.extraFiles
```
Comment (andreabedini) [inline Cabal-syntax/src/Distribution/PackageDescription/FieldGrammar.hs:146]:
```suggestion
    <*> monoidalFieldAla "extra-files" formatExtraSourceFiles L.extraFiles
    ^^^ availableSince CabalSpecV3_14 []
```

**Review by geekosaur — APPROVED** [review]

**Comment by geekosaur** [comment]
As noted by CI's tests, this changes the hashes of `GenericPackageDescription` and `LocalBuildInfo`, which will need to be updated.

File: Cabal-syntax/src/Distribution/PackageDescription/FieldGrammar.hs (hunk 1 above)
Comment (tbidne) [reply Cabal-syntax/src/Distribution/PackageDescription/FieldGrammar.hs:146]: Ah nice, thanks.

**Comment by tbidne** [comment]
Made the version change and updated the hashes (I got these locally from the validate script. I assume that or CI is fine).

Note that the [validate script](https://github.com/haskell/cabal/actions/runs/9773882232/job/26980951171?pr=10107) is failing:

```
rejecting: cabal-testsuite:setup.Cabal-3.13.0.0 (conflict: cabal-testsuite => cabal-testsuite:setup.Cabal>=3.10 && <3.11)
```

My guess is this is due to CI choosing the new cabal 3.12, and something needs to be updated (e.g. version bound, index-state).

**Comment by ulysses4ever** [comment]
@tbidne sorry about that. https://github.com/haskell/cabal/pull/10172 is supposed to fix it.

**Comment by ulysses4ever** [comment]
Please, rebase on `master` to receive the fix.

**Comment by ulysses4ever** [comment]
We need to decide on this. I'm not looking forward for bike-shedding the name and am happy with the current state.

File: changelog.d/issue-8817 [hunk 2]
```diff
@@ -0,0 +1,12 @@
+synopsis: Neutral field to add files to sdist
+packages: Cabal-syntax
```
Comment (ulysses4ever) [inline changelog.d/issue-8817:2]: This touches `Cabal` too, so it should be mentioned.

File: changelog.d/issue-8817 [hunk 3]
```diff
@@ -0,0 +1,12 @@
+synopsis: Neutral field to add files to sdist
+packages: Cabal-syntax
+prs: #10107
+issues: #8817
+significance: significant
+
+description: {
+  Adds the `extra-files` field to the cabal file specification. This is like
+  the other `extra-*` fields (e.g. `extra-source-files`, `extra-doc-files`)
+  exception it does _not_ add any other semantics (cf. `extra-source-files`
```
Comment (ulysses4ever) [inline changelog.d/issue-8817:10]:
```suggestion
  except it does _not_ add any other semantics (cf. `extra-source-files`
```

Also, it's be good to seize the opportunity to mention how extra-doc-files are special (they get copied by Haddock to the html directory).

Comment (tbidne) [reply changelog.d/issue-8817:10]: Do you want it in the issue? Or somewhere else (too)?

Comment (ulysses4ever) [reply changelog.d/issue-8817:10]: Only in this file. It will get to the top of release notes (due to `significance`), so, pretty visible…

Comment (tbidne) [reply changelog.d/issue-8817]: Got it, done.

**Comment by Mikolaj** [comment]
Yes, `extra-files` seems good enough and we need it now. Thanks for the PR!

File: changelog.d/issue-8817 (hunk 3 above)
Comment (ulysses4ever) [reply changelog.d/issue-8817]: Perfect, thank you!

**Review by ulysses4ever — APPROVED** [review]

**Comment by tbidne** [comment]
I am satisfied with the PR, so unless there is anything else, my understanding is that someone with the right permissions adds the `merge me` label.

**Comment by ulysses4ever** [comment]
Done. It will take several days for the bot to merge it.

**Pull request merged**